nemo/utils/app_state.py

```diff
@@ -0,0 +1,109 @@
+# -*- coding: utf-8 -*-
+# =============================================================================
+# Copyright (c) 2020 NVIDIA. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+# Sadly have to import this to avoid circular dependencies.
+import nemo
+from nemo.utils.metaclasses import Singleton
+
+
+class AppState(metaclass=Singleton):
```

Comment:
Do we need NeuralFactory if we have AppState? Alternatively, should we just let AppState create NeuralFactory?

Reply:
Once again: good catch! ... but this time this is for a purpose. :)

TLDR: In this PR I didn't wanted nor **brake** or **fix** its functionality - analogically to actions.py. Thus developed a second entity (Singleton) with a _clearly defined role_.

My problem with NF is... I do not fully understand the _role_ of Neural(Module)Factory. It seems to be an aggregation of everything, a shortcut to everything, plus something that manages global parameters, plus something that takes care of e.g. distributed training (so clearly is not "back-end independent"). I have asked @okuchaiev some time ago about some short presentation about his motivations and ideas behind that entity. After this, we can discuss what should be taken care by AppState and what should remain in NeuralFactory.

Reply:
For example, what NF is currently injecting into NMs __init__() doesn't make any sense.

Optimization parameters? Placement? Device? Why do we need that inside of NM? To move NMs to GPU during initialization...? But e.g. all DataLayers should NOT move data to GPU (if we want to use many DataLoader workers). So shouldn't it be the *action* who controls the CPU-GPU movement, including moving all modules to GPU?

This is just an example, but hope you got my point...

Reply:
`NeuralModuleFactory` used to be a factory. Currently, it is no longer a factory. But rather a "fake singleton" that keeps track of script status.

I suppose it is out of scope of this PR, but I think we should refactor NeuralModuleFactory soon after this PR.

Reply:
I don't understand why we need AppState if we have NeuralModuleFactory. Can't we move AppState's functionality into NeuralModuleFactory?

Reply:
I have explained that 21 hours ago. Nothing changed since then.

Asked you few weeks ago about a quick tutorial/explanation focused on NF, once we will be done with NG, we can focus on that issue if you want.

Reply:
Let me repeat, my conclusion is simple: **NF needs "rethinking"**. And doing that is out of the scope of this work.

Reply:
IMO, we only need one or the other. As it stands, AppState being an actual singleton behaves closer to what we want the current `NeuralModuleFactory` to do.

Reply:
Resolving for now, as we have agreed to discuss it later.